File: _archive/gather-shop/payment.py
"""Payment verification via Blockchair API.

Queries the BCH blockchain (via Blockchair) to verify that a transaction
exists and sent the correct amount to the shop's payment address.
Accepts 0-conf (mempool) transactions.
"""

import os
import re
from decimal import Decimal

from dotenv import load_dotenv
import httpx

from menu import get_item_price

load_dotenv()

_TX_ID_PATTERN = re.compile(r"^[0-9a-f]{64}$")

BLOCKCHAIR_URL = "https://api.blockchair.com/bitcoin-cash/dashboards/transaction"

# Shop's BCH receiving address. Loaded from .env or environment.
SHOP_BCH_ADDRESS = os.environ.get(
    "SHOP_BCH_ADDRESS",
    "YOUR_BCH_ADDRESS",
)

SATS_PER_BCH = 100_000_000


def _strip_prefix(address: str) -> str:
    """Strip 'bitcoincash:' prefix for comparison with Blockchair recipients."""
    return address.removeprefix("bitcoincash:")


async def verify_transaction(tx_id: str, expected_bch: Decimal) -> tuple[bool, str]:
    """Verify a BCH transaction via Blockchair.

    Checks that:
    1. The transaction exists on the blockchain (or mempool)
    2. At least one output goes to the shop's payment address
    3. That output's value >= expected amount

    Returns (ok, message).
    """
    if not _TX_ID_PATTERN.match(tx_id):
        return False, (
            f"Invalid transaction ID format. Expected a 64-character lowercase hex string. "
            f"Received: '{tx_id}' ({len(tx_id)} chars). "
            f"A valid tx_id looks like: b1e4b40416eb4f471ed66ee7c5fd5679cee39f38b7240660ad5e0db6bd854528"
        )

    expected_sats = int(expected_bch * SATS_PER_BCH)
    shop_addr = _strip_prefix(SHOP_BCH_ADDRESS)

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.get(f"{BLOCKCHAIR_URL}/{tx_id}")
    except httpx.HTTPError:
        return False, "Payment verification service unavailable. Please try again."

    if resp.status_code != 200:
        return False, "Payment verification service returned an error. Please try again."

    data = resp.json().get("data", {})

    # Blockchair returns data as {tx_hash: {...}} or empty dict/list if not found
    tx_data = data.get(tx_id) if isinstance(data, dict) else None
    if not tx_data:
        return False, f"Transaction {tx_id} not found on the BCH blockchain."

    outputs = tx_data.get("outputs", [])
    for output in outputs:
        if output.get("recipient") == shop_addr:
            if output.get("value", 0) >= expected_sats:
                return True, "Payment verified on blockchain."
            else:
                actual_bch = Decimal(output["value"]) / SATS_PER_BCH
                return False, (
                    f"Payment amount insufficient. "
                    f"Expected >= {expected_bch} BCH, found {actual_bch} BCH."
                )

    return False, (
        f"Transaction does not include payment to shop address ({SHOP_BCH_ADDRESS}). "
        f"Please send BCH to the payment_address returned in your order."
    )


def calculate_total(
    flavor: str, size: str, toppings: list[str]
) -> tuple[Decimal, None] | tuple[None, list[str]]:
    """Calculate order total from selected items.

    Returns (total, None) on success, or (None, list_of_invalid_ids) on failure.
    """
    invalid: list[str] = []

    flavor_price = get_item_price("flavors", flavor)
    if flavor_price is None:
        invalid.append(f"flavor '{flavor}' (see GET /menu/flavors)")

    size_price = get_item_price("sizes", size)
    if size_price is None:
        invalid.append(f"size '{size}' (see GET /menu/sizes)")

    bad_toppings = [t for t in toppings if get_item_price("toppings", t) is None]
    if bad_toppings:
        invalid.append(f"toppings {bad_toppings} (see GET /menu/toppings)")

    if invalid:
        return None, invalid

    total = Decimal(flavor_price) + Decimal(size_price)
    for topping_id in toppings:
        total += Decimal(get_item_price("toppings", topping_id))

    return total, None

File: gather-auth/go/shop/payment.go
package shop

import (
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"os"
	"regexp"
	"time"
)


var txIDPattern = regexp.MustCompile(`^[0-9a-f]{64}$`)

const blockchairURL = "https://api.blockchair.com/bitcoin-cash/dashboards/transaction"

var satPerBCH = big.NewInt(100_000_000)

func ShopBCHAddress() string {
	return os.Getenv("BCH_ADDRESS")
}

func stripPrefix(address string) string {
	if len(address) > 12 && address[:12] == "bitcoincash:" {
		return address[12:]
	}
	return address
}

// VerifyTransaction checks a BCH transaction via Blockchair.
// Returns (ok, message).
func VerifyTransaction(txID string, expectedBCH string) (bool, string) {
	if !txIDPattern.MatchString(txID) {
		return false, fmt.Sprintf(
			"Invalid transaction ID format. Expected a 64-character lowercase hex string. "+
				"Received: '%s' (%d chars).", txID, len(txID))
	}

	expectedRat := new(big.Rat)
	if _, ok := expectedRat.SetString(expectedBCH); !ok {
		return false, "Invalid expected amount."
	}
	// expectedSats = expectedBCH * 100_000_000
	expectedSats := new(big.Int)
	expectedSats.Mul(expectedRat.Num(), satPerBCH)
	expectedSats.Div(expectedSats, expectedRat.Denom())

	shopAddr := stripPrefix(ShopBCHAddress())

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(blockchairURL + "/" + txID)
	if err != nil {
		return false, "Payment verification service unavailable. Please try again."
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return false, "Payment verification service returned an error. Please try again."
	}

	var result struct {
		Data map[string]json.RawMessage `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return false, "Failed to parse blockchain response."
	}

	txRaw, ok := result.Data[txID]
	if !ok {
		return false, fmt.Sprintf("Transaction %s not found on the BCH blockchain.", txID)
	}

	var txData struct {
		Transaction struct {
			BlockID int64 `json:"block_id"`
		} `json:"transaction"`
		Outputs []struct {
			Recipient string `json:"recipient"`
			Value     int64  `json:"value"`
		} `json:"outputs"`
	}
	if err := json.Unmarshal(txRaw, &txData); err != nil {
		return false, "Failed to parse transaction data."
	}

	// Require at least 1 confirmation to prevent double-spend attacks
	if txData.Transaction.BlockID <= 0 {
		return false, "Transaction has 0 confirmations. Wait for at least 1 block confirmation and try again."
	}

	for _, out := range txData.Outputs {
		if out.Recipient == shopAddr {
			if big.NewInt(out.Value).Cmp(expectedSats) >= 0 {
				return true, "Payment verified on blockchain."
			}
			actualRat := new(big.Rat).SetFrac(big.NewInt(out.Value), satPerBCH)
			return false, fmt.Sprintf(
				"Payment amount insufficient. Expected >= %s BCH, found %s BCH.",
				expectedBCH, actualRat.FloatString(8))
		}
	}

	return false, fmt.Sprintf(
		"Transaction does not include payment to shop address (%s). "+
			"Please send BCH to the payment_address returned in your order.",
		ShopBCHAddress())
}

// VerifyDeposit checks a BCH deposit transaction and returns the actual amount
// sent to the platform address. Unlike VerifyTransaction, it doesn't require a
// specific expected amount — any amount is accepted.
// Returns (amountBCH, ok, message).
func VerifyDeposit(txID string) (string, bool, string) {
	if !txIDPattern.MatchString(txID) {
		return "", false, fmt.Sprintf(
			"Invalid transaction ID format. Expected a 64-character lowercase hex string. "+
				"Received: '%s' (%d chars).", txID, len(txID))
	}

	shopAddr := stripPrefix(ShopBCHAddress())

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(blockchairURL + "/" + txID)
	if err != nil {
		return "", false, "Payment verification service unavailable. Please try again."
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return "", false, "Payment verification service returned an error. Please try again."
	}

	var result struct {
		Data map[string]json.RawMessage `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", false, "Failed to parse blockchain response."
	}

	txRaw, ok := result.Data[txID]
	if !ok {
		return "", false, fmt.Sprintf("Transaction %s not found on the BCH blockchain.", txID)
	}

	var txData struct {
		Transaction struct {
			BlockID int64 `json:"block_id"`
		} `json:"transaction"`
		Outputs []struct {
			Recipient string `json:"recipient"`
			Value     int64  `json:"value"`
		} `json:"outputs"`
	}
	if err := json.Unmarshal(txRaw, &txData); err != nil {
		return "", false, "Failed to parse transaction data."
	}

	if txData.Transaction.BlockID <= 0 {
		return "", false, "Transaction has 0 confirmations. Wait for at least 1 block confirmation and try again."
	}

	// Sum all outputs to our address (agent might split across multiple outputs)
	totalSats := int64(0)
	for _, out := range txData.Outputs {
		if out.Recipient == shopAddr {
			totalSats += out.Value
		}
	}

	if totalSats == 0 {
		return "", false, fmt.Sprintf(
			"Transaction does not include payment to platform address (%s). "+
				"Send BCH to this address for deposits.",
			ShopBCHAddress())
	}

	amountRat := new(big.Rat).SetFrac(big.NewInt(totalSats), satPerBCH)
	return amountRat.FloatString(8), true, "Deposit verified on blockchain."
}

